subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java

```diff
@@ -241,6 +241,16 @@ public boolean isAllowOverlappingOutputs() {
             return true;
         }
 
+        @Override
+        public boolean isOverlappingOutputsDetected() {
+            return context.getOverlappingOutputs().isPresent();
+        }
+
+        @Override
+        public boolean isCleanupOutputsOnNonIncrementalExecution() {
+            return isRequiresInputChanges() && !isRequiresLegacyInputChanges();
+        }
```

Comment:
For some reason we have been using `is` getters on the `UnitOfWork`. I am happy to change the getters to `has` and `should` where it makes sense. We won't be able to use any shorthand notation from Groovy and Kotlin for the properties, though I think that is fine.

WDYT?

Reply:
I don't think we can decide this in the execution engine alone, since for example artifact transforms always want to do cleanup, even if they are not incremental.

Like we do it here, we can later on opt-in more tasks into automatic cleanup.

Reply:
> For some reason we have been using `is` getters on the `UnitOfWork`. I am happy to change the getters to `has` and `should` where it makes sense. We won't be able to use any shorthand notation from Groovy and Kotlin for the properties, though I think that is fine.

Let's do it!

> I don't think we can decide this in the execution engine alone, since for example artifact transforms always want to do cleanup, even if they are not incremental.

I forgot to remove that comment, sorry, it wasn't relevant.

Reply:
I did the renames.